design-proposals/klip-1-improve-udf-interfaces.md

```diff
@@ -0,0 +1,314 @@
+ensure that the validation does not unnecessarily prevent usage of `Struct`.
+
+A noteworthy extension that may be worth supporting is dedicated AVRO types in the signatures for 
+UDFs, but this is not covered in this KLIP.
+
+```java
+@Udf("Checks if the employee has a valid record (i.e. contains a valid name and email)")
+public boolean isValid(
+    @UdfParameter final Struct employee) {
+  return employee.get("firstname").matches("[A-Z][a-z]*")
+    && employee.get("email").endsWith("@company.io");
+}
+```
+
+There is more work to be done in order to support `Struct` as the return value of UDFs, namely we
+must have some mechanism to create the output schema. To address this, we can resolve the schema as 
+part of the UDF specification. This adds more structure and predictability, but may become tricky to
+evolve and we need a good API to do this, especially if it is necessary to specify complicated 
+nested structs. Below are three candidate ways to specify the schema:
+
+```java
+@UdfSchema
+public static final Schema SCHEMA = SchemaBuilder.struct()...build();
+
+@Udf("Checks if the employee has a valid record (i.e. contains a valid name and email)")
+@UdfReturn(value  = "STRUCT<'A' INT, 'B' VARCHAR>") // sample specification annotation
+@UdfReturn(file   = 'schema_def.kschema')           // another way pointing to a file
+@UdfReturn(schema = "name.space.MyClass.SCHEMA")    // another way that would resolve a java object
+public Struct generate() {
+  return new Struct(...);
+}
+```
+
+### Generics
+
+Today, we eagerly convert UDF output types to corresponding Connect schema types at the time that we
+compile the UDFs and not the time that we use them in a `SELECT` clause. Since there is no 
+corresponding "unresolved" type in Connect, we fail this conversion. To solve this problem, we can
+infer the type from the source schema and compile the code to work with any component type (e.g.
+omit the generics from the compiled code).
+
+```java
+@Udf("returns a sublist of 'list'")
+public <T> List<T> sublist(
+    @UdfParameter final List<T> list, 
+    @UdfParameter final int start, 
+    @UdfParameter final int end) {
+  return list.subList(start, end);
+}
+```
+
+We need to ensure that the output type can be inferred. This means that we need to fail if the 
+signature is something like `<T> List<T> convert(List<String> list)`. This can be done during 
+compilation by following a simple rule: any generic type used in the output must be present in at
+least one of the parameters. We can access this information via reflection.
+
+**NOTE:** Supporting a wildcard output type is not covered by this design since we would not be able 
+to generate the output schema for select statements, however supporting wildcards in the input types 
+(e.g. `Long length(List<?> list)`) should be possible.
+
+### Varargs
+
+Varargs boils down to supporting native Java arrays as parameters and ensuring that component types 
```

Comment:
How do we match struct varargs?

Reply:
I think solving your question

> How do we know how the schema of the struct parameter maps to a udf implementation, in the case a user wants to overload a function?

Would answer this as well. We make sure that each struct has the same schema as the one defined in the parameter schema.